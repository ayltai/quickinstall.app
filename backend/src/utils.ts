import { ErrorReporting, } from '@google-cloud/error-reporting';

const errors = new ErrorReporting();

export const handleError = (error : any) => {
    if (process.env.NODE_ENV === 'production') {
        errors.report(error);
    } else {
        console.error(error);
    }
};
