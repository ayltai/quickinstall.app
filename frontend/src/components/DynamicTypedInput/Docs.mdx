import { Controls, Description, Meta, Title, Primary, Stories, } from '@storybook/blocks';

import * as ComponentStories from './DynamicTypedInput.stories';

<Meta of={ComponentStories} />

<Title />

<Description />

# How to use dynamic typed inputs

Dynamic typed inputs are used when the required input data type is only known at runtime so that users can input data using an appropriate input field.

When using dynamic typed inputs,
* Use a label to explain exactly what data is required

# Default usage

<Primary />

# Parameters

<Controls sort='requiredFirst' />

<Stories
    title='Examples'
    includePrimary={false} />
