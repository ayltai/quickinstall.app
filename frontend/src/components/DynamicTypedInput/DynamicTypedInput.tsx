import Stack from '@mui/material/Stack';
import Switch from '@mui/material/Switch';
import TextField from '@mui/material/TextField';
import Typography from '@mui/material/Typography';
import React, { type FC, useCallback, useState, } from 'react';
import { useTranslation, } from 'react-i18next';

import type { DynamicTypedInputProps, } from './DynamicTypedInput.types';

/**
 * An input field that allows users to specify a boolean value, or input a number or string value depending on the type specified.
 * @param color Color of the input
 * @param title Title of the input
 * @param type Type of the input
 * @param value Value of the input
 * @param onChange Function to call when the input is changed
 * @param rest Other props
 */
export const DynamicTypedInput : FC<DynamicTypedInputProps> = ({
    color,
    title,
    type,
    value,
    onChange,
    ...rest
}) => {
    const [ currentValue, setCurrentValue, ] = useState(value);
    const [ error,        setError,        ] = useState(false);

    const { t, } = useTranslation();

    const handleChange = useCallback((event : React.ChangeEvent<HTMLInputElement>) => {
        setCurrentValue(event.target.value);

        if (onChange) {
            switch (type) {
                case 'boolean':
                    onChange(event.target.checked);

                    break;

                case 'number':
                    if (event.target.value && event.target.value.length > 0) {
                        if (isNaN(Number(event.target.value))) {
                            setError(true);
                        } else if (Number(event.target.value) >>> 0 === parseFloat(event.target.value)) {
                            setError(false);
                        } else {
                            setError(true);
                        }
                    } else {
                        setError(true);
                    }

                    onChange(Number(event.target.value));

                    break;

                case 'string':
                    if (event.target.value && event.target.value.length > 0) {
                        setError(false);
                    } else {
                        setError(true);
                    }

                    onChange(event.target.value);

                    break;

                default:
                    break;
            }
        }
    }, [ onChange, type, ]);

    return (
        <Stack
            spacing={1}
            display='flex'
            direction='row'
            alignItems='top'
            {...rest}>
            <Typography flexGrow={1}>
                {title}
            </Typography>
            {type === 'boolean' && (
                <Switch
                    checked={currentValue as boolean}
                    color={color}
                    onChange={handleChange} />
            )}
            {type === 'number' && (
                <TextField
                    required
                    color={color}
                    error={error}
                    helperText={error ? t('error.input.required') : undefined}
                    value={currentValue as number}
                    onChange={handleChange} />
            )}
            {type === 'string' && (
                <TextField
                    required
                    color={color}
                    error={error}
                    helperText={error ? t('error.input.required') : undefined}
                    value={currentValue as string}
                    onChange={handleChange} />
            )}
        </Stack>
    );
};
