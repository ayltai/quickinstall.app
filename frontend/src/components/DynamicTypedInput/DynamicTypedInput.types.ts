import type { HTMLAttributes, } from 'react';

export type DynamicTypedInputProps = Omit<HTMLAttributes<HTMLDivElement>, 'color' | 'onChange'> & {
    color?    : 'primary' | 'secondary',
    title     : string,
    type      : 'boolean' | 'number' | 'string',
    value     : boolean | number | string,
    onChange? : (value? : boolean | number | string) => void,
};
