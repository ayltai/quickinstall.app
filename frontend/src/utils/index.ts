import * as Sentry from '@sentry/react';
import { FirebaseApp, } from 'firebase/app';
import { getAnalytics, logEvent, } from 'firebase/analytics';

export const handleError = (error : any) => {
    if (import.meta.env.PROD) {
        Sentry.captureException(error);
    } else {
        console.error(error);
    }
};

export const logPageView = (app : FirebaseApp, {
    pageName,
    route,
} : {
    pageName : string,
    route    : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'page_view', {
            page_title    : pageName,
            page_location : route,
        });
    } catch (error) {
        handleError(error);
    }
};

export const logAddToCart = (app : FirebaseApp, {
    platform,
    itemId,
} : {
    platform : 'win' | 'mac',
    itemId   : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'add_to_cart', {
            items : [
                {
                    item_id : `${platform}/${itemId}`,
                },
            ],
        });
    } catch (error) {
        handleError(error);
    }
};

export const logRemoveFromCart = (app : FirebaseApp, {
    platform,
    itemId,
} : {
    platform : 'win' | 'mac',
    itemId   : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'remove_from_cart', {
            items : [
                {
                    item_id : `${platform}/${itemId}`,
                },
            ],
        });
    } catch (error) {
        handleError(error);
    }
};

export const logCheckout = (app : FirebaseApp, {
    platform,
    itemIds,
} : {
    platform : 'win' | 'mac',
    itemIds  : string[],
}) => {
    try {
        logEvent(getAnalytics(app), 'checkout', {
            items : itemIds.map((itemId : string) => ({
                item_id : `${platform}/${itemId}`,
            })),
        });
    } catch (error) {
        handleError(error);
    }
};

export const logSearch = (app : FirebaseApp, {
    platform,
    query,
} : {
    platform : 'win' | 'mac',
    query    : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'search', {
            search_term : `${platform}/${query}`,
        });
    } catch (error) {
        handleError(error);
    }
};

export const logShare = (app : FirebaseApp, {
    product,
    itemId,
} : {
    product : 'win/packages' | 'mac/packages' | 'tweaks',
    itemId  : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'share', {
            content_type : product,
            item_id      : itemId,
        });
    } catch (error) {
        handleError(error);
    }
};
