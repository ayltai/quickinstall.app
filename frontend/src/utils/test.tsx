import { createTheme, ThemeProvider, } from '@mui/material';
import { configureStore, type Store, } from '@reduxjs/toolkit';
import { render, } from '@testing-library/react';
import React, { type ReactNode, } from 'react';
import { Provider, } from 'react-redux';
import { MemoryRouter, } from 'react-router-dom';

import { appStoreApi, chocolateyApi, contentfulApi, homebrewApi, quickInstallApi, } from '../apis';
import { recipeReducer, preferenceReducer, spellReducer, } from '../states';
import { appTheme, } from '../styles';

export const createStore = (preloadedState? : unknown) => configureStore({
    preloadedState,
    reducer    : {
        recipe                          : recipeReducer,
        preference                      : preferenceReducer,
        spell                           : spellReducer,
        [ appStoreApi.reducerPath  ]    : appStoreApi.reducer,
        [ chocolateyApi.reducerPath ]   : chocolateyApi.reducer,
        [ contentfulApi.reducerPath ]   : contentfulApi.reducer,
        [ homebrewApi.reducerPath  ]    : homebrewApi.reducer,
        [ quickInstallApi.reducerPath ] : quickInstallApi.reducer,
    },
    middleware : getDefaultMiddleware => getDefaultMiddleware().concat(appStoreApi.middleware, chocolateyApi.middleware, contentfulApi.middleware, homebrewApi.middleware, quickInstallApi.middleware),
});

export const defaultStore = createStore();

export type RootState = ReturnType<typeof defaultStore.getState>;

// @ts-ignore
const theme = createTheme(appTheme);

const customRender = (ui : any, {
    preloadedState,
    store = createStore(preloadedState),
    ...rest
} : {
    preloadedState?   : Record<string, any>,
    store?            : Store<RootState>,
    [ rest : string ] : any,
} = {}) => render(ui, {
    wrapper : ({
        children,
    } : {
        children : ReactNode,
    }) => (
        <Provider store={store}>
            <MemoryRouter>
                <ThemeProvider theme={theme}>
                    {children}
                </ThemeProvider>
            </MemoryRouter>
        </Provider>
    ),
    ...rest,
});

export * from '@testing-library/react';

export { customRender as render, };
