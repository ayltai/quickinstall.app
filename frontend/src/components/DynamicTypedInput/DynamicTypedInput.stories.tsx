import type { Meta, StoryObj, } from '@storybook/react';

import { DynamicTypedInput, } from './DynamicTypedInput';

const meta : Meta<typeof DynamicTypedInput> = {
    component : DynamicTypedInput,
    title     : 'Components/Inputs/Dynamic Typed Input',
    argTypes  : {
        title : {
            description : 'The title of the component',
        },
        type  : {
            description : 'The type of the component',
        },
        value : {
            description : 'The value of the component',
        },
    },
};

export default meta;

type Story = StoryObj<typeof DynamicTypedInput>;

export const Default : Story = {
    args : {
        title : 'Dynamic Typed Input',
        type  : 'boolean',
        value : true,
    },
};

export const BooleanType : Story = {
    args : {
        title : 'Dynamic Typed Input',
        type  : 'boolean',
        value : true,
    },
};

export const NumberType : Story = {
    args : {
        title : 'Dynamic Typed Input',
        type  : 'number',
        value : '1',
    },
};

export const StringType : Story = {
    args : {
        title : 'Dynamic Typed Input',
        type  : 'string',
        value : 'String value',
    },
};
