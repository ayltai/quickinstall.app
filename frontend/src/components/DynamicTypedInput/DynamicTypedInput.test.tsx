import React from 'react';

import { fireEvent, render, } from '../../utils/test';
import { DynamicTypedInput, } from './DynamicTypedInput';

describe('<DynamicTypedInput />', () => {
    it('renders boolean typed input correctly', () => expect(render(
        <DynamicTypedInput
            title='Title'
            type='boolean'
            value={true} />
    )).toMatchSnapshot());

    it('renders string typed input correctly', () => expect(render(
        <DynamicTypedInput
            title='Title'
            type='string'
            value='test' />
    )).toMatchSnapshot());

    it('triggers onChange when the boolean input is changed', () => {
        const handleChange = vi.fn();

        const { getByRole, } = render(
            <DynamicTypedInput
                title='Title'
                type='boolean'
                value={true}
                onChange={handleChange} />
        );

        fireEvent.click(getByRole('checkbox'));

        expect(handleChange).toHaveBeenCalled();
        expect(handleChange).toHaveBeenCalledWith(false);
    });

    it('triggers onChange when the string input is changed', () => {
        const handleChange = vi.fn();

        const { getByRole, } = render(
            <DynamicTypedInput
                title='Title'
                type='string'
                value='test'
                onChange={handleChange} />
        );

        fireEvent.change(getByRole('textbox'), {
            target: {
                value: 'Dummy',
            },
        });

        expect(handleChange).toHaveBeenCalled();
        expect(handleChange).toHaveBeenCalledWith('Dummy');
    });
});
