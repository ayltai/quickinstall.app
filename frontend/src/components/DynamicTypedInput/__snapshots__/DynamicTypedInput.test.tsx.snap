// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<DynamicTypedInput /> > renders boolean typed input correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiStack-root css-crjfvx-MuiStack-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-1lpz5xa-MuiTypography-root"
        >
          Title
        </p>
        <span
          class="MuiSwitch-root MuiSwitch-sizeMedium css-julti5-MuiSwitch-root"
        >
          <span
            class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary Mui-checked PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary Mui-checked Mui-checked css-1tb1e9q-MuiButtonBase-root-MuiSwitch-switchBase"
          >
            <input
              checked=""
              class="PrivateSwitchBase-input MuiSwitch-input css-1m9pwf3"
              type="checkbox"
            />
            <span
              class="MuiSwitch-thumb css-jsexje-MuiSwitch-thumb"
            />
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiSwitch-track css-1yjjitx-MuiSwitch-track"
          />
        </span>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiStack-root css-crjfvx-MuiStack-root"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-1lpz5xa-MuiTypography-root"
      >
        Title
      </p>
      <span
        class="MuiSwitch-root MuiSwitch-sizeMedium css-julti5-MuiSwitch-root"
      >
        <span
          class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary Mui-checked PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary Mui-checked Mui-checked css-1tb1e9q-MuiButtonBase-root-MuiSwitch-switchBase"
        >
          <input
            checked=""
            class="PrivateSwitchBase-input MuiSwitch-input css-1m9pwf3"
            type="checkbox"
          />
          <span
            class="MuiSwitch-thumb css-jsexje-MuiSwitch-thumb"
          />
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiSwitch-track css-1yjjitx-MuiSwitch-track"
        />
      </span>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<DynamicTypedInput /> > renders string typed input correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiStack-root css-crjfvx-MuiStack-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-1lpz5xa-MuiTypography-root"
        >
          Title
        </p>
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-1dyv133-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r0:"
              required=""
              type="text"
              value="test"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-ihdtdm"
              >
                <span
                  class="notranslate"
                >
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiStack-root css-crjfvx-MuiStack-root"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-1lpz5xa-MuiTypography-root"
      >
        Title
      </p>
      <div
        class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-1dyv133-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r0:"
            required=""
            type="text"
            value="test"
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-ihdtdm"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
